File: KthSmallestEleinBST.java
// Time Complexity : O(k)- as we traverse till we reach the Kth smallest element
// Space Complexity : O(N) - Internal stack
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    int count = 0;
    int result = 0;
    public int kthSmallest(TreeNode root, int k) {
       count = k;
       element(root);
       return result;
    }
    public void element(TreeNode root){
        if(root.left != null){
            element(root.left);
        }
        count--;
        if(count == 0){
            result = root.val;
            return;
        }
        if(root.right != null){
            element(root.right);
        }
    }
}

// Your code here along with comments explaining your approach

File: KthSmallestEleInBST.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
from typing import Optional


class Solution:
    def kthSmallest(self, root: Optional[TreeNode], k: int) -> int:
        if root is None:
            return 

        self.count = k
        self.result = None

        def inorder(node):
            if node is None:
                return
            if self.result == None:
                inorder(node.left)
            self.count -= 1
            if self.count == 0:
                self.result = node
            if self.result == None:
                inorder(node.right)
        
        inorder(root)
        return self.result.val

